File: codes/leetcode/PathSum.cpp
/*
************************************************
username    :   smmehrab
fullname    :   s.m.mehrabul islam
email       :   mehrab.24csedu.001@gmail.com
institute   :   university of dhaka, bangladesh
session     :   2017-2018
************************************************
*/
class Solution {
private:
    bool solve(TreeNode* node, int sum) {
        bool hasPath = false;
        hasPath = (!node->left && !node->right) ? (sum==node->val) : hasPath;
        hasPath = (node->left) ? hasPath || solve(node->left, sum-(node->val)) : hasPath;
        hasPath = (node->right) ? hasPath || solve(node->right, sum-(node->val)) : hasPath;
        return hasPath;
    }
    
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if(!root) return false;
        return solve(root, sum);
    }
};

File: codes/leetcode/PathSum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):

    def traverse(self, node, sum):
        has_path = False
        if (not node.left and not node.right):
            has_path = (sum==node.val)
        if node.left:
            has_path |= self.traverse(node.left, (sum-node.val))
        if node.right:
            has_path |= self.traverse(node.right, (sum-node.val))
        return has_path

    def hasPathSum(self, root, targetSum):
        """
        :type root: TreeNode
        :type targetSum: int
        :rtype: bool
        """
        if not root:
            return False
        return self.traverse(root, targetSum)
